src/workerd/jsg/filesystem.c++

```diff
@@ -0,0 +1,537 @@
+#include "filesystem.h"
+
+#include "buffersource.h"
+
+namespace workerd::jsg::fs {
+
+namespace {
+constexpr size_t kMaxReadLength = 32 * 1024;
+
+void validatePathWithNoSeparators(kj::StringPtr path) {
+  try {
+    auto parsed = kj::Path::parse(path);
+    JSG_REQUIRE(parsed.size() == 1, Error, "Invalid path: \"", path, "\"");
+  } catch (kj::Exception& e) {
+    JSG_FAIL_REQUIRE(Error, "Invalid path: \"", path, "\"");
+  }
+}
+
+template <bool Writable = false>
+class DirectoryBase final: public Directory {
+ public:
+  DirectoryBase() = default;
+  DirectoryBase(kj::HashMap<kj::String, kj::OneOf<kj::Own<File>, kj::Own<Directory>>> entries)
+      : entries(kj::mv(entries)) {
+    KJ_DASSERT(!Writable);
+  }
+
+  const Stat stat() const override {
+    return Stat{
+      .type = Node::Type::DIRECTORY,
+      .size = 0,
+      .lastModified = lastModified,
+      .writable = Writable,
+    };
+  }
+
+  size_t count(Lock& js, kj::Maybe<Type> typeFilter = kj::none) const override {
+    KJ_IF_SOME(type, typeFilter) {
+      size_t counter = 0;
+      forEach(js, [&counter](Lock&, kj::StringPtr name, const Stat stat) {
+        counter++;
+        return true;
+      }, type);
+      return counter;
+    }
+    return entries.size();
+  }
+
+  void forEach(Lock& js,
+      kj::FunctionParam<bool(Lock&, kj::StringPtr, const Stat stat)> fn,
+      kj::Maybe<Type> typeFilter = kj::none) const override {
+    for (auto& entry: entries) {
+      KJ_SWITCH_ONEOF(entry.value) {
+        KJ_CASE_ONEOF(file, kj::Own<File>) {
+          KJ_IF_SOME(type, typeFilter) {
+            if (type != Node::Type::FILE) continue;
+          }
+          if (!fn(js, entry.key, file->stat())) {
+            break;
+          }
+        }
+        KJ_CASE_ONEOF(dir, kj::Own<Directory>) {
+          KJ_IF_SOME(type, typeFilter) {
+            if (type != Node::Type::DIRECTORY) continue;
+          }
+          if (!fn(js, entry.key, dir->stat())) {
+            break;
+          }
+        }
+      }
+    }
+  }
+
+  kj::Maybe<kj::OneOf<kj::Own<File>, kj::Own<Directory>>> tryOpen(
+      Lock& js, kj::PathPtr path, kj::Maybe<Type> createAs) override {
+    if (path.size() == 0) {
+      return kj::Maybe<kj::OneOf<kj::Own<File>, kj::Own<Directory>>>(addRef(js));
+    }
+    KJ_IF_SOME(found, entries.find(path[0])) {
+      path = path.slice(1, path.size());
+      if (path.size() == 0) {
+        // We found the entry, return it.
+        KJ_SWITCH_ONEOF(found) {
+          KJ_CASE_ONEOF(file, kj::Own<File>) {
+            return kj::Maybe<kj::OneOf<kj::Own<File>, kj::Own<Directory>>>(file->addRef(js));
+          }
+          KJ_CASE_ONEOF(dir, kj::Own<Directory>) {
+            return kj::Maybe<kj::OneOf<kj::Own<File>, kj::Own<Directory>>>(dir->addRef(js));
+          }
```

Comment:
Since they have similar APIs, these lines can be reduced into 1, and be simplified.

Reply:
Not really. Since we're using the `kj::OneOf` and these are separate types, they'll need separate branches.